user_src/data_handler.cxx

```diff
@@ -0,0 +1,155 @@
+void data_handler()
+{
+    // We keep receiving and processing messages until terminate is true and no more frames are buffered
+    while (!(terminate && frame_buffer.empty()))
+    {
+        // Load the next can frame into frame;
+        get_next_frame();
+
+        // Convert the frame into a can message
+        extract_message();
+
+        // Then either send the message to the server or store it on the SD Card
+        handle_message();
+    }
+
+    // There will be no more messages to process and all frames read by the SocketCAN Reader have been processed
+
+    printf(PRINT_HEADER "Terminated DataHandler successfully!");
+}
+
+// Retrieves the next can_frame from the frame_buffer and stores it into frame
+static void get_next_frame()
+{
+    // Enter critical section
+    // Lock the mutex
+    can_lock.lock();
+    // Wait for the frame_buffer queue to be non-empty
+    while (frame_buffer.empty())
+    {
+        if(terminate){ // If we need to terminate and the frame_buffer is empty, stop the thread
+            return;
+        }
+        CAN_FRAME_BUFFER_EMPTY.wait(can_lock); // What if we were already waiting and then terminate becomes true and no more frames come in.
+    }
+
+    // Receive the frame at the front of the queue
+    frame = frame_buffer.front();
+    // Remove the frame in the front from the queue
+    frame_buffer.pop();
+    // Signal the SocketCAN Reader that the queue is no longer full
+    CAN_FRAME_BUFFER_FULL.notify_all();
+    // Unlock the mutex
+    can_lock.unlock();
+    // Exit the critical section
+}
+
+// Extract data from the can_frame into a string which contains only the necessary data from the can_frame and adds a timestamp.
+static void extract_message()
+{
+    // Create the approximate timestamp at which this CAN-message was received
+    uint64_t timestamp = duration_cast<seconds>(system_clock::now().time_since_epoch()).count();
+
+    // The actual id of the can_frame, which is either the first 11 or 29 bits based on spec
+    uint32_t actual_id;
+
+    // Check the CAN spec/version of this message, which is determined by the last bit of can_id
+    if (frame.can_id & (1 << 31))
+    { // If the 31st bit is set, then we have spec 2.0B = 29 bit id
+        actual_id = frame.can_id & ((1 << 29) - 1);
+    }
+    else
+    { // If the 31st bit is not set, then we have spec 2.0A = 11 bit id
+        actual_id = frame.can_id & ((1 << 11) - 1);
+    }
+
+    char test[] = "%lx"; // ??
+
+    // Convert the extracted data alongside the timestamp into a string
+    char extract[31];
+    // 10?13? + 1 + 8 + 1 + 16 = 36/9 bytes
+    int nbytes = sprintf(extract, "%llu#%08lx#%x%x%x%x%x%x%x%x", timestamp, frame.can_id, frame.data[0], 
+    frame.data[1], frame.data[2], frame.data[3], frame.data[4], frame.data[5], frame.data[6], frame.data[7]);
+
+    if(nbytes != 40){
+        printf(PRINT_HEADER "Warning: Extracted can_frame not equal to the expected 40 bytes!");
```

Comment:
What happens in the subsequent parts of the program if such a frame (where nbytes != 40) is pushed to the database/sd controller? Is this something that can realistically happen and if so wouldn't you want to discard such a "broken" frame?